### PR TITLE
Deploy cere smart contract

### PR DESCRIPTION
**Pull request by karthikeyan-cere**

### CODE REVIEW

File: scripts/tokenomics-emulation/cere-smart-contract.ts [hunk 1]
```diff
@@ -62,11 +68,45 @@ class CereSmartContract {
...
+    return new Promise((res, rej) => {
+      tx.signAndSend(
+        sender,
+        Network.sendStatusCb.bind(this, res, rej)
+      ).catch((err) => rej(err));
+    });
+  }
+
+  /**
+   * Deploy the code on chain
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/cere-smart-contract.ts:92]: You have method `deploy` but comment `Deploy code` is here. Please, fix it.

File: scripts/tokenomics-emulation/cere-smart-contract.ts [hunk 2]
```diff
@@ -62,11 +68,45 @@ class CereSmartContract {
...
+      .transfer({ value, gasLimit }, destination, amount, 0)
       .paymentInfo(sender);
     return txnFee;
   }
+
+  public async deploy(sender: KeyringPair) {
+    console.log(`Deploy smart contract`);
+    const code = new CodePromise(this.api, cere02Abi, cere01Wasm);
+
+    const params = [10000000000, ""];
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/cere-smart-contract.ts:81]: What is `100000`. Move to config.json

Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/cere-smart-contract.ts:81]: What is `""`. Move to config.json

File: scripts/tokenomics-emulation/config.json [hunk 3]
```diff
@@ -83,6 +83,14 @@
         "name": "existential-deposit-transfer",
         "amount": 1
       },
+      {
+        "name": "deploy-cere-smart-contract",
+        "code_hash": "0x343f41b8cf072966ce537da45c0a9ca6a95db97dcd4e076c8a6fab99e7742023",
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/config.json:88]: What is it? We can not hardcode any hashes.

File: scripts/tokenomics-emulation/cere-smart-contract.ts (hunk 1 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts:94]: Done

File: scripts/tokenomics-emulation/cere-smart-contract.ts (hunk 2 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts:83]: Removed

Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts:83]: Removed

File: scripts/tokenomics-emulation/config.json (hunk 3 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/config.json:88]: These are code_hash to deploy the smart contract

File: scripts/tokenomics-emulation/cere-smart-contract.ts [hunk 4]
```diff
@@ -62,11 +70,101 @@ class CereSmartContract {
...
+    const unsub = await blueprint.tx.new(
+      endowment,
+      gasLimit,
+      initialValue,
+      dsAccounts
+    );
+    
+    return new Promise((res, rej) => {
+      unsub
+        .signAndSend(sender, this.sendStatusCb.bind(this, res, rej))
```
[the remark this answers is not in the document]
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts:122]: Updated the code update the smart contract address in the config.json

File: scripts/tokenomics-emulation/ddc-smart-contract.ts [hunk 5]
```diff
@@ -62,6 +68,116 @@ class DdcSmartContract {
...
+      tier2ThroughtputLimit,
+      tier2StorageLimit,
+      tier3Limit,
+      tier3ThroughtputLimit,
+      tier3StorageLimit,
+      symbol
+    );
+    return new Promise((res, rej) => {
+      unsub
+        .signAndSend(sender, this.sendStatusCb.bind(this, res, rej))
```
[the remark this answers is not in the document]
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/ddc-smart-contract.ts:129]: Updated the code update the smart contract address in the config.json

File: scripts/tokenomics-emulation/cere-smart-contract.ts [hunk 6]
```diff
@@ -62,11 +68,76 @@ class CereSmartContract {
...
+  /**
+   * Process the events to set smart contract address
+   * @param events Events
+   */
+  private async handleEvents(events) {
+    console.log("Handling Events");
+    events.forEach((event) => {
+      if (event.event.data.length === 2) {
+        configFile.network.cere_sc_address = event.event.data[1].toString();
+        fs.writeFileSync("config.json", JSON.stringify(configFile));
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/cere-smart-contract.ts:136]: Will it be updated in `config` variable which already in use in the script?

File: scripts/tokenomics-emulation/config.json [hunk 7]
```diff
@@ -83,6 +83,30 @@
...
+        "tier_1_fee": 10,
+        "tier_1_throughput_limit": 12,
+        "tier_1_storage_limit":14,
+        "tier_2_fee": 20,
+        "tier_2_throughput_limit": 22,
+        "tier_2_storage_limit":24,
+        "tier_3_fee": 30,
+        "tier_3_throughput_limit": 32,
+        "tier_3_storage_limit":34,
+        "symbol": "KART"
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/config.json:108]: :-)

File: scripts/tokenomics-emulation/ddc-smart-contract.ts [hunk 8]
```diff
@@ -62,6 +67,82 @@ class DdcSmartContract {
...
+
+  /**
+   * Process the events to set smart contract address
+   * @param events Events
+   */
+   private async handleEvents(events) {
+    console.log("Handling Events");
+    events.forEach((event) => {
+      if (event.event.data.length === 2) {
+        configFile.network.ddc_sc_address = event.event.data[1].toString();
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/ddc-smart-contract.ts:140]: Will it be updated in config variable which already in use in the script?

File: scripts/tokenomics-emulation/cere-smart-contract.ts (hunk 6 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts:136]: Yes.

File: scripts/tokenomics-emulation/ddc-smart-contract.ts (hunk 8 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/ddc-smart-contract.ts:140]: yes.

File: scripts/tokenomics-emulation/cere-smart-contract.ts [hunk 9]
```diff
@@ -62,11 +68,76 @@ class CereSmartContract {
...
+      ).catch((err) => rej(err));
+    });
+  }
+
+  /**
+   * Deploy the blueprint on chain
+   * @param sender smart contract owner
+   * @returns Transaction hash
+   */
+  public async bluePrint(
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/cere-smart-contract.ts:100]: What to do with bluePrint?

File: scripts/tokenomics-emulation/ddc-smart-contract.ts [hunk 10]
```diff
@@ -62,6 +67,82 @@ class DdcSmartContract {
...
+      ).catch((err) => rej(err));
+    });
+  }
+
+  /**
+   * Deploy the blluprint on chain
+   * @param sender smart contract owner
+   * @returns Transaction hash
+   */
+  public async bluePrint(
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/ddc-smart-contract.ts:93]: What to do with bluePrint?

**Review by AndreiNavoichyk — APPROVED** [review]

File: scripts/tokenomics-emulation/cere-smart-contract.ts (hunk 9 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts]: renamed to `deployBlueprint`

File: scripts/tokenomics-emulation/ddc-smart-contract.ts (hunk 10 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/ddc-smart-contract.ts]: Renamed to `deployBlueprint`

**Pull request merged**